clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -86,33 +86,32 @@ public FutureRecordMetadata tryAppend(long timestamp, byte[] key, byte[] value,
      * Complete the request
      * 
      * @param baseOffset The base offset of the messages assigned by the server
-     * @param timestamp The timestamp returned by the broker.
+     * @param responseTimestamp The timestamp returned by the broker.
      * @param exception The exception that occurred (or null if the request was successful)
      */
-    public void done(long baseOffset, long timestamp, RuntimeException exception) {
+    public void done(long baseOffset, long responseTimestamp, RuntimeException exception) {
         log.trace("Produced messages to topic-partition {} with base offset offset {} and error: {}.",
                   topicPartition,
                   baseOffset,
                   exception);
+
+        // set the future before invoking the callbacks as we rely on its state for the `onCompletion` call
+        produceFuture.set(topicPartition, baseOffset, responseTimestamp, exception);
+
         // execute callbacks
         for (Thunk thunk : thunks) {
             try {
                 if (exception == null) {
-                    // If the timestamp returned by server is NoTimestamp, that means CreateTime is used. Otherwise LogAppendTime is used.
-                    RecordMetadata metadata = new RecordMetadata(this.topicPartition,  baseOffset, thunk.future.relativeOffset(),
-                                                                 timestamp == Record.NO_TIMESTAMP ? thunk.future.timestamp() : timestamp,
-                                                                 thunk.future.checksum(),
-                                                                 thunk.future.serializedKeySize(),
-                                                                 thunk.future.serializedValueSize());
+                    RecordMetadata metadata = thunk.future.value();
```

Comment:
This works fine for the callbacks, but doesn't work for the future returned by the `send` method without additional bookkeeping. We pass `ProduceRequestResult` into `FutureRecordMetdata` when we build it so that we can update the futures returned to users without having to store them ourselves. It's good not to store them as they can be GC'd sooner in case users discard them.